### DISCUSSION TITLE
Java API: Passing a collection as generic varargs

### DISCUSSION THREAD

**Discussion by zhangwen0411** [Z3Prover/z3#5645] [Q&A] [answered]
I have a `FuncDecl` and want to apply it to some arguments stored in an `ArrayList<Expr<?>>`. The signature of `FuncDecl.apply` looks like:
```
public Expr<R> apply(Expr<?> ... args)
```
but Java doesn't allow me to create an `Expr<?>[]` from my `ArrayList<Expr<?>>`. I could create an `Expr[]`, but then the Java compiler warns me about "Raw use of parameterized class 'Expr'".

Is there a good way to call `apply` on a collection of expressions?

**Comment by wintersteiger** [comment, marked as answer]
I haven't touched an Java stuff in a while, buy what about `ArrayList::toArray()`?

**Reply by zhangwen0411** [reply]
Yes! I was able to do:
```
Expr<?>[] args = arrayList.toArray(new Expr[0]);
```
and then write `funcDecl.apply(args)` without the compiler complaining. Thank you!
